tools/ingest_historical_csv.py

```diff
@@ -104,16 +111,15 @@ async def _main(args):
     log.info('invoking ingestion for {} CSV files'.format(len(args.csv)))
     func, url = _get_executor(args)
     sem = asyncio.BoundedSemaphore(args.concurrency)
-    async with ClientSession() as session:
-        tasks = []
-        for csv in args.csv:
-            log.debug('Convert records in {} to JSON'.format(csv))
-            df = pd.read_csv(csv, index_col=None, header=0)  # pylint: disable=invalid-name
-            log.debug('CSV Record count {count}'.format(count=len(df)))
-            updated_df = _update_df(df, csv)
-            task = asyncio.ensure_future(func(df=updated_df, session=session, url=url, csv=csv, sem=sem))
-            tasks.append(task)
-        _ = await asyncio.gather(*tasks)
+    for csv in args.csv:
+        log.debug('Convert records in {} to JSON'.format(csv))
+        df = pd.read_csv(csv, index_col=None, header=0)  # pylint: disable=invalid-name
+        log.debug('CSV Record count {count}'.format(count=len(df)))
+        updated_df = _update_df(df)
+        # Runnning one file at a time to overcome duplicate issue for similar record across different ecosystem
+        # though all data inside dataframe/file will be inserted parallel
+        await func(df=updated_df, url=url, csv=csv, sem=sem)
```

Comment:
Problem has to be fixed in the backend side If async io causes duplicate in the graph. Disabling asyncio seems to be a workaround and it defeats the purpose of running replicas in the backend.

Reply:
I would suggest combine all csv into single data frame, remove the duplicates and do the insertion in parallel using asyncio.

Reply:
That is what we already discussed earlier when we discussed "Concurrent request analysis". This insert will happen through backend through either this script or through inference pipeline where we are not expecting that much load and decided to run sequentially.

Reply:
> Problem has to be fixed in the backend side If async io causes duplicate in the graph. Disabling asyncio seems to be a workaround and it defeats the purpose of running replicas in the backend.

That is what we already discussed earlier when we discussed "Concurrent request analysis". This insert will happen through backend through either this script or through inference pipeline where we are not expecting that much load and decided to run sequentially.

Reply:
> I would suggest combine all csv into single data frame, remove the duplicates and do the insertion in parallel using asyncio.

That also I believe we can't do, as it will result into duplicate records 
Ex : if one pcve is with the two ecosystem then after combining we will get data more than one row for same pcve url, like below example. 

> ecosystem url...... 
----------
openshift  https://github.com/org/pkg/pull/1....
knative  https://github.com/org/pkg/pull/1.....
kubevirt  https://github.com/org/pkg/pull/1.....

And if we call all record together with asyncio then we may have duplicate record. So can't do much here also Arun :)

Reply:
@rajusem, in that case can't we group the rows by url instead of ecosystem and use single asyncio task to insert all events related to that specific url?

If we are not going to use any form of concurrency, then keeping asycio as a dependency doesn't make any sense. Better remove asycio and use requests/urlib, which would simplify the code.

Reply:
We are using similar stuff, in place of doing group by url and use single asyncio task, we are inserting all data in a single file with single asyncio task. It's not like we are not using at all, as I mentioned earlier, we are using it to load data parallel from a single file.

Reply:
@rajusem , IIUC, all records with in a single file is inserted concurrently. Thats fine.

Actually we decided to go with a single CSV file for all [ecosystems](https://github.com/kubesecurity/osa-data-collector/blob/276bd3e0853eb76dda7bec87336bcc9a6a4256aa/src/bq_data_collector.py#L222) right? In that case our assumption would break and we will endup with duplicate data.

Reply:
@arajkumar  : As we are not using data collector as of now and we are getting data from inference pipeline where we are getting 3 different file. And in future when we will have datacollector integrated with inference pipeline then we should get **single url** but ecosystem should come as **"openshift,knative,kubevirt"** so no question of getting duplicate data, and need to handle the same in API accordingly in insertion. Once we have that flow, will come will work/modify accordingly.